File: frontend/src/services/templates.js
/**
 * Visibility and validation for garment templates.
 *
 * The exact mirror of core/templates.py. If you change a rule here, change it
 * there in the same commit: the browser hiding a field the server still demands
 * is how a form saves an order with the wrong measurements on it.
 *
 * A hidden field is never required and is never stored.
 */

const asList = (value) => (Array.isArray(value) ? value : [value]);

const isEmpty = (value) =>
  value === null ||
  value === undefined ||
  value === '' ||
  (Array.isArray(value) && value.length === 0);

export function evaluateRule(rule, values) {
  if (!rule) return true;
  if (rule.all) return rule.all.every((r) => evaluateRule(r, values));
  if (rule.any) return rule.any.some((r) => evaluateRule(r, values));

  const actual = values[rule.field];
  const op = rule.op || 'eq';
  const expected = rule.value;

  if (op === 'is_set') return !isEmpty(actual);
  // A multiselect answer is a list, so 'in' means "any chosen value matches" --
  // picking Fall + Pico must reveal both the fall and the pico field.
  if (op === 'in') {
    const chosen = new Set(asList(actual));
    return asList(expected).some((v) => chosen.has(v));
  }
  if (op === 'not_in') {
    const chosen = new Set(asList(actual));
    return !asList(expected).some((v) => chosen.has(v));
  }
  if (op === 'neq') return actual !== expected;
  return actual === expected;
}

export const isVisible = (field, values) => evaluateRule(field.visible_when, values);

/** Every field of a section that currently applies, in render order. */
export function visibleFields(section, values) {
  if (!section) return [];
  return section.fields.filter((field) => isVisible(field, values));
}

export const getSection = (template, key) =>
  template?.sections?.find((section) => section.key === key);

/**
 * Drop the answers to fields that have become hidden.
 *
 * Called on every change: without it, choosing Peplum, typing a flare length and
 * switching to Corset leaves the flare length behind, and the cutter sees a
 * measurement for a panel that is not being made.
 */
export function pruneHidden(template, values) {
  const kept = {};
  (template?.sections || []).forEach((section) => {
    section.fields.forEach((field) => {
      if (isVisible(field, values) && values[field.key] !== undefined) {
        kept[field.key] = values[field.key];
      }
    });
  });
  return kept;
}

/** Returns {field_key: message}; empty when the spec is good. */
export function validateSpec(template, values, { partial = false } = {}) {
  const errors = {};

  (template?.sections || []).forEach((section) => {
    section.fields.forEach((field) => {
      if (!isVisible(field, values)) return;

      const raw = values[field.key];
      if (isEmpty(raw)) {
        if (field.is_required && !partial) {
          errors[field.key] = `${field.label} is required.`;
        }
        return;
      }

      const rules = field.validation || {};
      if (field.field_type === 'number') {
        const number = Number(raw);
        if (Number.isNaN(number)) {
          errors[field.key] = `${field.label} must be a number.`;
        } else if (rules.min !== undefined && number < rules.min) {
          errors[field.key] = `${field.label} cannot be below ${rules.min}.`;
        } else if (rules.max !== undefined && number > rules.max) {
          errors[field.key] = `${field.label} cannot exceed ${rules.max}.`;
        }
      } else if (field.field_type === 'select' || field.field_type === 'multiselect') {
        const allowed = new Set((field.options || []).map((o) => o.value));
        const unknown = asList(raw).filter((v) => !allowed.has(v));
        if (unknown.length) {
          errors[field.key] = `${field.label}: unknown option ${unknown.join(', ')}.`;
        }
      } else if (rules.max_length && String(raw).length > rules.max_length) {
        errors[field.key] = `${field.label} is limited to ${rules.max_length} characters.`;
      }
    });
  });

  const { trial_date: trial, delivery_date: delivery } = values;
  if (trial && delivery && trial > delivery) {
    errors.trial_date = 'Trial date must fall before the delivery date.';
  }

  return errors;
}

/** Split a flat spec into the two columns GarmentJob stores it in. */
export function splitSpec(template, values) {
  const measurementKeys = new Set(
    (getSection(template, 'measurements')?.fields || []).map((f) => f.key)
  );
  const spec = {};
  const measurements = {};
  Object.entries(values).forEach(([key, value]) => {
    (measurementKeys.has(key) ? measurements : spec)[key] = value;
  });
  return { spec, measurements };
}

File: core/templates.py

from decimal import Decimal, InvalidOperation


class SpecValidationError(ValueError):


    def __init__(self, errors):
        self.errors = errors
        super().__init__(f"{len(errors)} invalid field(s): {', '.join(sorted(errors))}")



def _as_list(value):
    return value if isinstance(value, (list, tuple)) else [value]


def evaluate_rule(rule, values):

    if not rule:
        return True
    if 'all' in rule:
        return all(evaluate_rule(r, values) for r in rule['all'])
    if 'any' in rule:
        return any(evaluate_rule(r, values) for r in rule['any'])

    actual = values.get(rule['field'])
    op = rule.get('op', 'eq')
    expected = rule.get('value')

    if op == 'is_set':
        return actual not in (None, '', [], {})
    if op == 'in':
        return bool(set(_as_list(actual)) & set(_as_list(expected)))
    if op == 'not_in':
        return not (set(_as_list(actual)) & set(_as_list(expected)))
    if op == 'neq':
        return actual != expected
    return actual == expected


def is_visible(field, values):
    return evaluate_rule(field.visible_when, values)


def visible_fields(template, values):

    result = []
    for section in template.sections.all():
        for field in section.fields.all():
            if is_visible(field, values):
                result.append(field)
    return result



def _clean_number(field, raw, errors):
    try:
        number = Decimal(str(raw))
    except (InvalidOperation, TypeError, ValueError):
        errors[field.key] = f"{field.label} must be a number."
        return None

    rules = field.validation or {}
    minimum, maximum = rules.get('min'), rules.get('max')
    if minimum is not None and number < Decimal(str(minimum)):
        errors[field.key] = f"{field.label} cannot be below {minimum}."
    elif maximum is not None and number > Decimal(str(maximum)):
        errors[field.key] = f"{field.label} cannot exceed {maximum}."
    return str(number)


def _clean_choice(field, raw, errors, multi):
    allowed = {o.value for o in field.options.all() if o.is_active}
    chosen = _as_list(raw) if multi else [raw]
    unknown = [c for c in chosen if c not in allowed]
    if unknown:
        errors[field.key] = f"{field.label}: unknown option {', '.join(map(str, unknown))}."
        return None
    return list(chosen) if multi else raw


def validate_spec(template, spec, *, partial=False):
    errors = {}
    cleaned = {}
    known = {}

    for section in template.sections.all():
        for field in section.fields.all():
            known[field.key] = field

    unknown_keys = set(spec) - set(known)
    if unknown_keys:
        for key in unknown_keys:
            errors[key] = "Unknown field for this garment."

    for key, field in known.items():
        if not is_visible(field, spec):
            continue  # hidden: not required, not stored

        raw = spec.get(key)
        if raw in (None, '', [], {}):
            if field.is_required and not partial:
                errors[key] = f"{field.label} is required."
            continue

        field_type = field.field_type
        if field_type == 'number':
            value = _clean_number(field, raw, errors)
        elif field_type == 'select':
            value = _clean_choice(field, raw, errors, multi=False)
        elif field_type == 'multiselect':
            value = _clean_choice(field, raw, errors, multi=True)
        elif field_type == 'boolean':
            value = bool(raw)
        else:
            value = raw
            max_length = (field.validation or {}).get('max_length')
            if max_length and isinstance(raw, str) and len(raw) > max_length:
                errors[key] = f"{field.label} is limited to {max_length} characters."

        if key not in errors and value is not None:
            cleaned[key] = value

    _validate_cross_field(cleaned, errors)

    if errors:
        raise SpecValidationError(errors)
    return cleaned


def _validate_cross_field(cleaned, errors):

    trial, delivery = cleaned.get('trial_date'), cleaned.get('delivery_date')
    if trial and delivery and str(trial) > str(delivery):
        errors['trial_date'] = "Trial date must fall before the delivery date."
